WeatherPlannerDemo/Network/BaseApiClient.swift

```diff
@@ -0,0 +1,149 @@
+//
+//  BaseApiClient.swift
+//  WeatherPlannerDemo
+//
+//  Created by Alen Sebalj on 07.12.2020..
+//
+
+import Foundation
+
+public class BaseApiClient: ApiClientProtocol {
+    
+    private let baseUrl: String
+    private let urlSession: URLSession
+    
+    public init(baseUrl: String, urlSession: URLSession) {
+        self.baseUrl = baseUrl
+        self.urlSession = urlSession
+    }
+    
+    public func get<ResultType: Decodable>(
+        path: String,
+        queryParameters: [String : String]?,
+        memberType: ResultType.Type,
+        resultHandler: @escaping (Result<ResultType, ApiError>) -> ()
+    ) {
+        executeAndReturn(path: path, method: .get, parameters: queryParameters, resultHandler: resultHandler)
+    }
+    
+    private func executeAndReturn<ParamsType: Encodable, ResultType: Decodable>(
+        path: String,
+        method: HTTPMethod,
+        parameters: ParamsType? = nil,
+        resultHandler: @escaping (Result<ResultType, ApiError>) -> ()
+    ) {
+        let request = buildRequest(path: path, method: method, parameters: parameters)
+
+        let urlTask = urlSession.dataTask(with: request) {  (data, response, error) in
+            guard
+                let httpReponse = response as? HTTPURLResponse,
+                //when [weak self], self if always nil ???
+                //let self = self,
+                let statusCode = HttpStatusCode(rawValue: httpReponse.statusCode)
+            else { return }
+            
+            if let error = self.mapToApiError(status: statusCode){
+                print("Request failed with error: \(error)")
+                resultHandler(.failure(error))
+                return
+            }
+            
+            guard let result: ResultType = self.parse(data: data) else {
```

Comment:
This is a memory leak. You are using strong self reference here. You need to weakify self. This is commented above.